README

Net::DNS - Perl DNS Resolver Module
===================================


   TABLE OF CONTENTS
   -----------------

1.  Description
2.  Availability
3.  Prerequisites
4.  Installation
5.  Running Tests
6.  Demonstration Scripts
7.  Dynamic Updates
8.  Signed Queries & Updates
9.  DNSSEC
10. Bugs
11. Copyright
12. License
13. Staying Tuned
14. Acknowledgments


1. DESCRIPTION
   -----------

Net::DNS is a DNS resolver implemented in Perl. It allows the programmer
to perform nearly any type of DNS query from a Perl script. For details
and examples, please read the Net::DNS manual page.

To read about the latest features, see the Changes file. To find out about
known bugs and to see what is planned for future versions, see the CPAN RT
ticket list.

The author invites feedback on Net::DNS. If there is something you would
like to have added, please let me know. If you find a bug, please send me
the information described in the BUGS section below.

See http://www.net-dns.org/blog/ for announcements about Net::DNS.


2. AVAILABILITY
   ------------

You can get the latest version of Net::DNS from the Comprehensive Perl
Archive Network (CPAN) or from the module's homepage:

	https://metacpan.org/release/Net-DNS

or through

	http://www.net-dns.org/

Additionally a subversion repository is made available through

	http://www.net-dns.org/svn/net-dns/

The version on the "trunk" (http://www.net-dns.org/svn/net-dns/trunk) is
the version that is targeted for next release.

Please note that the SVN version at any given moment may be broken.


3. PREREQUISITES
   -------------

The availability of prerequisites for Net::DNS is tested at installation
time. These are the core packages that need to be available:

	Digest::HMAC
	Digest::MD5  
	Digest::SHA  
	File::Spec      
	IO::Socket      
	IO::Socket::IP
	MIME::Base64     
	Time::Local
	Test::More       

The availability of these optional packages is tested at runtime:

	Digest::BubbleBabble
	Net::DNS::SEC
	Net::LibIDN2

You can obtain the latest version of Perl from:

	http://www.cpan.org

Some of the demonstration and contributed scripts may require additional
modules -- see demo/README and contrib/README for details.

Note that the Test::More module is actually part of the Test-Simple
distribution. See the FAQ (lib/Net/DNS/FAQ.pod) for more information.


4. INSTALLATION
   ------------

Please install any modules mentioned in the PREREQUISITES section above.
If you do not, some features of Net::DNS will not work. When you run "perl
Makefile.PL", Perl should complain if any of the required modules is
missing.

To build this module, run the following commands:

	tar xvzf Net-DNS-?.??.tar.gz
	cd  Net-DNS-?.??
	perl Makefile.PL
	make
	make test
	make install

If you do not wish to run the online tests, the '--no-online-tests' option
can be used. Similarly, '--online-tests' will enable the online tests.
Online tests will be run by default, but the result will not adversely
affect the outcome of test suite.

Also, if you do not wish to run the IPv6 tests, the '--no-IPv6-tests'
option can be used. Similarly, '--IPv6-tests' will enable the IPv6 tests.


5. RUNNING TESTS
   -------------

If any of the tests fail, please contact the author with the output from
the following command:

	make test TEST_VERBOSE=1


6. DEMONSTRATION SCRIPTS
   ---------------------

There are a few demonstration scripts in the demo/ directory -- see
demo/README for more information. Contributed scripts are in the contrib/
directory -- see contrib/README.

The author would be happy to include any contributed scripts in future
versions of this module. All I ask is that they be documented (preferably
using POD) and that the contributor's name and contact information be
mentioned somewhere.


7. DYNAMIC UPDATES
   ---------------

Net::DNS supports DNS dynamic updates as documented in RFC 2136; for more
information and examples, please see the Net::DNS::Update manual page.

Here is a summary of the update semantics for those interested (see RFC
2136 for details):

    PREREQUISITE SECTION
                   # RRs   NAME   TTL   CLASS   TYPE   RDLENGTH   RDATA
                   -----   ----   ---   -----   ----   --------   -----
    yxrrset          1     name    0     ANY    type       0      empty
    yxrrset          1+    name    0    class   type   rdlength   rdata
    nxrrset          1     name    0     NONE   type       0      empty
    yxdomain         1     name    0     ANY     ANY       0      empty
    nxdomain         1     name    0     NONE    ANY       0      empty


    UPDATE SECTION
                   # RRs   NAME   TTL   CLASS   TYPE   RDLENGTH   RDATA
                   -----   ----   ---   -----   ----   --------   -----
    add RRs          1+    name   ttl   class   type   rdlength   rdata
    del RRset        1     name    0     ANY    type       0      empty
    del all RRsets   1     name    0     ANY     ANY       0      empty
    del RRs          1+    name    0     NONE   type   rdlength   rdata


8. SIGNED QUERIES & UPDATES
   ------------------------

Net::DNS supports the TSIG resource record to perform signed queries and
updates (RFC 2845). See the Net::DNS::Packet and Net::DNS::Update manual
pages for examples.

If you're using the BIND nameserver, the BIND FAQ shows how to generate
keys and configure the nameserver to use them:

	http://www.nominum.com/resources/faqs/bind-faq.html

TSIG support is new and isn't yet complete. Please use with caution on
production systems. Feedback on TSIG functionality would be most welcome.


9. DNSSEC
   ------

The extensions to enable the DNSSEC signature generation and verification
functions are distributed separately as Net::DNS::SEC. The package is
available from CPAN.


10. BUGS
    ----

Net::DNS, although begun in 1997, is still under development and may still
contain a few bugs. Please see CPAN RT and Changes file for more
information.

We recommend that you exercise caution when using Net::DNS to maintain a
production nameserver via dynamic updates. Always test your code
*thoroughly*. The Net::DNS authors accept no blame if you corrupt your
zone. That warning in place, We are aware of one large company that has
used Net::DNS to make thousands of dynamic updates per day for at least
three years without any problems.

Please use the following form to submit bug reports:

	https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-DNS

If you find any bugs, please report each in a separate "rt.cpan.org"
report along with the following information:

	* subject field containing a concise descriptive summary
	* version of Perl (output of 'perl -V' is best)
	* version of Net::DNS
	* operating system type and version
	* version of nameserver (if known)
	* exact text of error message or description of problem
	* the shortest possible program that exhibits the problem
	* the specific queries you are making, if the fault can be
	  demonstrated using Internet nameservers

If we do not have access to a system similar to yours, you may be asked
to insert some debugging lines and report back on the results. The more
focussed the help and information you provide, the better.

Net::DNS is currently maintained at NLnet Labs (www.nlnetlabs.nl) by:
	Willem Toorop.

Between 2005 and 2012 Net::DNS was maintained by:
	Olaf Kolkman and his team.

Between 2002 and 2004 Net::DNS was maintained by Chris Reinhardt.

Net::DNS was created in 1997 by Michael Fuhr.


11. COPYRIGHT
    ---------

Authorship of individual components and significant contributions is shown
in the copyright notice attached to the relevant documentation. Copyright
in all components is retained by their respective authors.


12. LICENSE
    -------

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted, provided
that the original copyright notices appear in all copies and that both
copyright notice and this permission notice appear in supporting
documentation, and that the name of the author not be used in advertising
or publicity pertaining to distribution of the software without specific
prior written permission.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.


13. STAYING TUNED
    -------------

http://www.net-dns.org is a web site dedicated to the development of
Net::DNS. Announcements about Net::DNS and Net::DNS::SEC will be done
through the Net::DNS weblog at http://www.net-dns.org/blog/. An RSS feed
for the weblog is available.

If you want to have access to the latest and greatest code a subversion
repository is made available through

	http://www.net-dns.org/svn/net-dns/

The version on the "trunk" (http://www.net-dns.org/svn/net-dns/trunk) is
the version that is targeted for next release.

Please note that code from the SVN repositories trunk and development
branches may be broken at any time.


14. ACKNOWLEDGMENTS
    ---------------

Thanks to Mike for letting me take care of his baby.

Thanks to Chris for maintaining Net::DNS for a couple of years.

Thanks to Olaf for maintaining Net::DNS for over eight years.

Thanks to Rob Brown and Dick Franks for all their patches and input.

Thanks to all who have used Net::DNS and reported bugs, made suggestions,
contributed code, and encouraged me to add certain features. Many of these
people are mentioned by name in the Changes file; lack of mention should
be considered an oversight and not a conscious act of omission.

Thanks to Larry Wall and all who have made Perl possible.

Thanks to Paul Albitz and Cricket Liu for allowing me [OK: that is Mike]
to write the Net::DNS section in the programming chapter of DNS and BIND,
3rd Edition. This chapter in earlier editions was very helpful while I
was developing Net::DNS, and I was proud to contribute to it.

Thanks to Paul Vixie and all who have worked on the BIND nameserver, which
I've used exclusively while developing Net::DNS.

Thanks to Andreas Gustafsson for DNAME support, and for all the work he
has done on BIND 9.

Olaf acknowledges the RIPE NCC for allowing Net::DNS maintenance to take
place as part of 'the job'.

Thanks to the team that maintains wireshark. Without its marvelous
interface, debugging of bugs in wireformat would be so much more
difficult.

Thanks to the thousands who participate in the open-source community. I
have always developed Net::DNS using open-source systems and I am proud
to make Net::DNS freely available to the world.


---- $Id$



Changes
$Id$                       -*-text-*-


**** 1.36 [unrleased]

	Adopt JSON as presentation notation for EDNS options.
	Zero packet->id not allowed in outbound packet.
	Remove deprecated 2-argument TSIG->create() method.
	Revise TSIG test scripts and documentation.


**** 1.35 Oct 4, 2022

	Improve SVCB error reporting.

Fix rt.cpan.org #144328
	accept_reply test fails with matched consecutive "random"
	generated packet->id

Fix rt.cpan.org #144299
	Spelling errors.


**** 1.34 May 30, 2022

	Improve robustness of EDNS option compose/decompose functions.
	Simplify code in Makefile.PL.

Fix rt.cpan.org #142426
	Avoid "Useless use of a constant in void context" warning.


**** 1.33 Dec 16, 2021

Fix rt.cpan.org #137768
	Test t/05-SVCB.t on Perl 5.18.0 fails with deep recursion.

Fix rt.cpan.org #144136/#132921
	$resolver->send wrongly overwrites RD flag in user's packet.


**** 1.32 Jul 16, 2021

	Text: Offer both Unicode and escaped-ASCII strings.
	Add LICENSE file to comply with Fedora/RedHat announcement.

Fix rt.cpan.org #136666
	Net::DNS::RR::ZoneFile parser erroneously strips line
	terminators in quoted string forming part of multiline RR.


**** 1.31 May 2, 2021

	Improve implementation of SVCB record.


**** 1.30 Mar 30, 2021

	Simplify parsing of multi-line RRs in zone file.
	Improve robustness of "dry" resolver tests.
	Avoid deep recursion in non-fatal test report.


**** 1.29 Nov 18, 2020

	Include test number in summary of failed non-fatal tests.
	Remove Net::DNS::SEC specific tests.
	Fix faulty test plan in t/08-recurse.t.


**** 1.28 Oct 23, 2020

	Eliminate indirect object syntax.
	Eliminate grep/map <expression>.


**** 1.27 Sep 11, 2020

Fix rt.cpan.org #133203
	Net::DNS::RR::LOC erroneously strips non default values from
	string representation


**** 1.26 Aug 6, 2020

	Add HTTPS/SVCB packages.

Fix rt.cpan.org #132921
	EDNS OPT handling


**** 1.25 Jun 26, 2020

	Parsing of TSIG keyfiles made more robust.


**** 1.24 May 27, 2020

	Accept TSIG key generated by BIND tsig-keygen.
	Add Net::DNS::RR::AMTRELAY package.


**** 1.23 Mar 18, 2020

	Deprecate 2-argument form of TSIG create().

Fix rt.cpan.org #132170
	[Documentation] Problems with TSIG on ddns update.

Fix rt.cpan.org #131906
	Undefined errorstring/warning when axfr fails


**** 1.22 Feb 13, 2020

Fix rt.cpan.org #131579
	Parse issue in Net::DNS::RR->token

Feature
	Provide rudimentary decode and print for DSO packet.


**** 1.21 Aug 30, 2019

	Fix error report for non-existent or recursive zone file $INCLUDE.
	Emit one deprecation warning on invocation of obsolete method.
	Rework OPT.pm EDNS0 option construction.
	Remove obsolete Net::DNS::RR::DLV package.
	Add Net::DNS::RR::ZONEMD package.

Fix rt.cpan.org #128901
	background TCP query logic expects to read entire response at once


**** 1.20 Mar 22, 2019

	TSIG MAC representation changed to Base64 (align with BIND).
	Update Parameters.pm to resync with IANA registry.
	Refactor resolver test scripts.
	Revise documentation examples to use AAAA instead of A records.

Fix rt.cpan.org #128081
	Recurse.pm fails to resolve domain "kickboxingireland.ie"

Fix rt.cpan.org #127307
	Provide a more informative exception report if application code
	has no "use Net::DNS::SEC" declaration but nevertheless attempts
	to invoke the DNSSEC sign or verify features.


**** 1.19 Nov 14, 2018

	Show structure of EDNS options using Perl-like syntax.

Fix rt.cpan.org #127557
	Net::DNS::Resolver::Base should use 3 args open

Fix rt.cpan.org #127182
	Incorrect logic can cause DNS search to emit fruitless queries.


**** 1.18 Sep 21, 2018

	Documentation revised to remove ambigous use of "answer" which
	has been used to refer to both the answer section of a packet
	and the entire reply packet received from a nameserver.

Fix rt.cpan.org #127018
	Net::DNS::ZoneFile->parse() fails if include directory specified.

Fix rt.cpan.org #127012
	DNS resolution broken when options ndots used in /etc/resolv.conf


**** 1.17 Jul 25, 2018

Fix rt.cpan.org #125890
	AXFR: 1 record per packet responses.

Fix rt.cpan.org #125889
	New NSEC3 for empty non-terminal leaves type bitmap undefined.

Fix rt.cpan.org #125882
	RDATA name compression pointer calculated incorrectly.


**** 1.16 Jul 15, 2018

Feature
	New NSEC3 encloser(), nextcloser() and wildcard() instance
	methods return closest encloser, "next closer" and putative
	wildcard names respectively.

Feature
	Add new NSEC covers() instance method.

Feature
	New NSEC typemap() instance method interrogates type list.

	IO::Socket::INET6 removed from recommended module metadata.
	IPv6 requires IO::Socket::IP which is now a core package.
	No requirement to escape @ in unquoted contiguous string.


**** 1.15 Feb 9, 2018

	GOST R 34.11-94 hash algorithm: end of life 1st Jan 2018
	per sunset clause in successor standard GOST R 34.11-2012.
	Digest::GOST removed from the recommended module metadata,
	but will still be used if available.


**** 1.14 Dec 15, 2017

Fix rt.cpan.org #123702
	'use base' should not be used in packages with several
	subpackages defined

Fix rt.cpan.org #123676
	Net::DNS::Nameserver malformed message on big axfr


**** 1.13 Oct 18, 2017

Feature IDN query support
	Queries for domain names containing non-ASCII characters are
	now possible on Unicode platforms using CPAN Net::LibIDN2


**** 1.12 Aug 18, 2017

Fix rt.cpan.org #122586
	Persistent UDP reports false timeouts

Fix rt.cpan.org #122352
	bgsend(): TCP retry can stall for IO::Socket::IP before 0.38

Feature
	CDS / CDNSKEY: Implement RFC8078 erratum 5049.


**** 1.11 Jun 26, 2017

Fix rt.cpan.org #122138
	Send a UDP query with udppacketsize=512

Feature
	Extract default resolver configuration from OS/390 MVS datasets.
	Thanks to Sandra Carroll and Yaroslav Kuzmin for their assistance.


**** 1.10 May 5, 2017

Fix rt.cpan.org #120748
	Net::DNS::Resolver::MSWin32 critical issue
	Thanks to Dmytro Zagashev for his valuable assistance during

Makefile.PL
#
# $Id$	-*-perl-*-
#

use 5.008009;
use strict;
use warnings;
use ExtUtils::MakeMaker;

my $distro = 'Net::DNS';
my $module = join '/', 'lib', split /::/, "$distro.pm";
my $author = ['Dick Franks', 'Olaf Kolkman', 'Michael Fuhr'];
$author = join ', ', @$author if $ExtUtils::MakeMaker::VERSION < 6.58;


# See perldoc ExtUtils::MakeMaker for details of how to influence
# the contents of the Makefile that is written.

my %metadata = (
	NAME		   => $distro,
	VERSION_FROM	   => $module,
	ABSTRACT_FROM	   => $module,
	AUTHOR		   => $author,
	LICENSE		   => 'mit',
	MIN_PERL_VERSION   => '5.008009',
	CONFIGURE_REQUIRES => {
		'ExtUtils::MakeMaker' => 6.48,
		'Getopt::Long'	      => 2.43,
		'IO::File'	      => 1.14,
		'IO::Socket::IP'      => 0.38,
		},
	TEST_REQUIRES => {
		'ExtUtils::MakeMaker' => 0,
		'File::Find'	      => 1.13,
		'File::Spec'	      => 3.29,
		'IO::File'	      => 1.14,
		'Test::Builder'	      => 0.80,
		'Test::More'	      => 0.80,
		} );


my %platform = (			## platform-specific dependencies
	MSWin32 => {
		'Win32::IPHelper'    => 0.07,
		'Win32::API'	     => 0.55,
		'Win32::TieRegistry' => 0.24,
		} );

my $platform = $platform{$^O} || {};


my %prerequisite = (
	'Carp'		 => 1.10,
	'Digest::HMAC'	 => 1.03,
	'Digest::MD5'	 => 2.13,
	'Digest::SHA'	 => 5.23,
	'Encode'	 => 2.26,
	'Exporter'	 => 5.63,
	'File::Spec'	 => 3.29,
	'IO::File'	 => 1.14,
	'IO::Select'	 => 1.17,
	'IO::Socket'	 => 1.30,
	'IO::Socket::IP' => 0.38,
	'MIME::Base64'	 => 2.13,
	'PerlIO'	 => 1.05,
	'Scalar::Util'	 => 1.19,
	'Time::Local'	 => 1.19,
	%$platform
	);


my %optional = (
	'Digest::BubbleBabble' => 0.02,
	'Net::LibIDN2'	       => 1.00,
	);


## IMPORTANT:	THE USE AND/OR HANDLING OF STRONG ENCRYPTION TECHNOLOGIES IS
##		PROHIBITED OR SEVERELY RESTRICTED IN MANY TERRITORIES.
delete $prerequisite{'Net::DNS::SEC'};	## MUST NOT list Net::DNS::SEC dependency in metadata.
delete $optional{'Net::DNS::SEC'};	## Require explicit user action to install Net::DNS::SEC.


my @debris = qw(.resolv.conf *.lock);


#
# Get the command line args
#
use constant USE_GETOPT => defined eval { require Getopt::Long };

my $help	 = 0;
my $IPv6_tests	 = 1;
my $online_tests = 2;			##  2 = non-fatal, 1 = on, 0 = off  ##

my @options = (
	'online-tests!'		 => \$online_tests,
	'noonline-tests'	 => sub { $online_tests = 0; },
	'non-fatal-online-tests' => sub { $online_tests = 2; },
	'IPv6-tests!'		 => \$IPv6_tests,
	'help!'			 => \$help
	);


if ( USE_GETOPT && !Getopt::Long::GetOptions(@options) ) {
	print "Error: Unrecognized option.\n";
	print "Try perl Makefile.PL --help for more information\n";
	exit 1;
}


if ($help) {
	print <<'EOT';

Usage: perl Makefile.PL [OPTION...]

Prepare Makefile for building and installing Net::DNS

  --online-tests		Perform tests by communicating with the
				outside world.	Beware that their success is
				also dependent on outside influences.

  --noonline-tests		Skip online tests completely.

  --IPv6-tests			Perform IPv6 specific online tests. Default is
				the same as regular online tests.

  --noIPv6-tests		Skip IPv6 specific online test

  --non-fatal-online-tests	Perform online test, but do not let failures 
				negatively affect the outcome of running
				make test.  This is the default.

EOT
	exit 0;
}


#
# Enable tests if we have internet connection	(code lifted from LWP)
#
use constant USE_SOCKET_IP => defined eval 'use IO::Socket::IP 0.38; 1;';    ## no critic
require IO::Socket::INET unless USE_SOCKET_IP;

if ($online_tests) {
	my $class = USE_SOCKET_IP ? 'IO::Socket::IP' : 'IO::Socket::INET';

	my $socket = $class->new(
		PeerAddr => 'dns.google.',
		PeerPort => 53,		## check 53/TCP not blocked ##
		Timeout	 => 10
		);

	unless ($socket) {
		$online_tests = 0;
		print <<"EOT";

Unable to establish TCP/IP connection to the global Internet.
$@

Disabling online tests altogether...

EOT
	}
}


# set up online testing configuration files.
my $enable   = 't/online.enabled';
my $IPv6     = 't/IPv6.enabled';
my $nonfatal = 't/online.nonfatal';
push @debris, $enable, $IPv6, $nonfatal;

if ($online_tests) {
	require IO::File;
	my $fh1 = IO::File->new( $enable, '>' ) or die "Can't touch $enable $!";
	close($fh1);

	if ( $online_tests == 2 ) {
		print "\nActivating Non Fatal Online Tests...\n";
		my $fh2 = IO::File->new( $nonfatal, '>' ) or die "Can't touch $nonfatal $!";
		close($fh2);
	} else {
		print "\nActivating Online Tests...\n";
	}

	if ( USE_SOCKET_IP && $IPv6_tests ) {
		print "\nActivating IPv6 Tests...\n";
		my $fh3 = IO::File->new( $IPv6, '>' ) or die "Can't touch $IPv6 $!";
		close($fh3);
	}

	print <<'EOT';

Warning!
========
Online tests depend on conditions beyond the control of Net::DNS. The tests 
check for the expected results when both Net::DNS and the outside world are
functioning properly. In case of failure it is often undecidable if the error
lies within Net::DNS or elsewhere.

EOT
}


WriteMakefile(				## Makefile & distribution metadata
	%metadata,
	PREREQ_PM  => {%prerequisite},
	META_MERGE => {recommends => {%optional}},
	clean	   => {FILES	  => "@debris"},
	);

exit;


package MY;				## customise generated Makefile

sub test {
	return shift->SUPER::test() if $^O =~ /cygwin|MSWin/i;

	return join '', shift->SUPER::test(), <<'END';
# suppress parallel test execution
FULLPERLRUN = HARNESS_OPTIONS=j1:c $(FULLPERL)
END
}


sub dist {
	return join '', shift->SUPER::dist(), <<'END';

# $(PERM_RWX) raises security issues downstream
PREOP = $(CHMOD) $(PERM_RW) $(DISTVNAME)$(DFSEP)contrib$(DFSEP)* $(DISTVNAME)$(DFSEP)demo$(DFSEP)*
END
}


sub install {
	my $self = shift;

	my %install_type = qw(perl INSTALLPRIVLIB site INSTALLSITELIB vendor INSTALLVENDORLIB);
	my $install_site = join '', '$(DESTDIR)$(', $install_type{$self->{INSTALLDIRS}}, ')';
	for ($install_site) {
		s/\$\(([A-Z_]+)\)/$self->{$1}/eg while /\$\(/;	# expand Makefile macros
		s|([/])[/]+|$1|g;				# remove gratuitous //s
	}

	eval "require $distro";		## no critic
	my @version = ( 'version', eval { $distro->VERSION } );

	my $nameregex = join '\W+', '', split /::/, "$distro.pm\$";
	my @installed = grep { $_ && m/$nameregex/io } values %INC;
	my %occluded;

	foreach (@installed) {
		my $path = m/^(.+)$nameregex/io ? $1 : '';
		my %seen;
		foreach (@INC) {
			$seen{$_}++;				# find $path in @INC
			last if $_ eq $path;
		}
		foreach ( grep { !$seen{$_} } @INC ) {
			$occluded{$_}++;			# suppress install
		}
	}

	return $self->SUPER::install(@_) unless $occluded{$install_site};

	my $message;
	warn $message = <<"AMEN";
##
##	The install location for this version of $distro differs
##	from the existing @version in your perl library at
##	@installed
##
##	The installation would be rendered ineffective because the
##	existing @version occurs in the library search path before
##	$install_site
##
##	The generated Makefile supports build and test only.
##
AMEN
	my $echo = '	$(NOECHO) $(ECHO) "##"';
	$message =~ s/##/$echo/eg;

	return join '', <<"END";
install :
	$message
	\$(NOECHO) \$(FALSE)
END
}


sub postamble {
	return <<'END';
test_cover :
	cover -delete
	HARNESS_PERL_SWITCHES=-MDevel::Cover $(MAKE) test
	cover
END
}


__END__

